### DISCUSSION TITLE
Using Lingui in monorepo over several projects

### DISCUSSION THREAD

**Discussion by danielkcz** [lingui/js-lingui#908] [Q&A] [answered]
We are trying to set up some sort of robust architecture that should be about composing different projects from other small pieces - modules within a single monorepo.

It's all configured based on TypeScript path aliases instead of packages that would need to be published to NPM. It works nicely so far. It looks roughly like this.

```json
    "baseUrl": "src",
    "paths": {
      "~tools/*": ["../../tools/src/*"]
    },
```

### Reusable I18n module

Ideally, there should be a separate module (still within monorepo) that consolidates all i18n logic including a single version of Lingui. All other modules will simply use this module.

The problem I am currently solving is with macros. When I do simply this.

```js
export { t } from '@lingui/macro'

// then import in the module code
import { t } from '~tools/lingui'
// tried also relative path in case aliasing is an issue
import { t } from '../../tools/src/lingui'
```

It works syntactically, but the extraction process silently ignores it and nothing is extracted. No output is even generated to the temp folder as I tried to debug it. It's a bit strange why Babel compiler stays silent there. When I change it to direct `@lingui/macro` it works, so setup is generally correct.

I expect this is one of many problems for this approach (ie path aliases), but let's start with basics.

**Comment by tricoder42** [comment, marked as answer]
You can't use macros from modules with invalid names. See official [`babel-plugin-macros`](https://github.com/kentcdodds/babel-plugin-macros/blob/master/other/docs/author.md#filename) documentation for a list of valid macro names.

Second, even if you solved it, you would have to update `@lingui/babel-plugin-extract-messages` plugin to actually see your module as well. right now it only looks for `@lingui/macro` and `@lingui/react`.

**Reply by danielkcz** [reply]
Hmm, I thought that `@lingui/babel-plugin-extract-messages` is no longer required and it's all handled by macro? I guess I was wrong, that's unfortunate.

Anyway, I was afraid there will be more hurdles to it. Would it make sense to have this configurable on Lingui side? I am not sure honestly how macros work in this matter and if it's even possible. I could certainly try to contribute if you think there is a way toward this.

**Reply by tricoder42** [reply]
I don't think you could configure it on Lingui side — macros are executed by `babel-plugin-macros` so you should definitely start there.

**Reply by danielkcz** [reply]
Ok thanks, this is probably not worth the hassle ultimately. It will have to keep using `@lingui/macro` installed in monorepo root and other runtime logic in the common module.
